File: allcode/401-500/440findKthNumber.py
# 给定整数n和k，返回[1, n]中字典序第k小的数字。
#
#
#
# 示例 1:
#
# 输入: n = 13, k = 2
# 输出: 10
# 解释: 字典序的排列是 [1, 10, 11, 12, 13, 2, 3, 4, 5, 6, 7, 8, 9]，所以第二小的数字是 10。
# 示例 2:
#
# 输入: n = 1, k = 1
# 输出: 1
#
#
# 提示:
#
# 1 <= k <= n <= 109


class Solution:
    def findKthNumber(self, n: int, k: int) -> int:

        def helper(s, k):
            for i in range(10):
                if int(s) * 10 + i <= n:
                    helper(s + str(i), k - 1)
                    break
        s = str(n)
        N = len(s)
        s1 = 1
        s2 = 0
        s3 = []
        for i in range(N):
            s1 = '1' * (N - i)
            s3.append(int(s1) * int(s[i]))
        print(s3)
        print(s2)


so = Solution()
print(so.findKthNumber(121,1))
#print(so.diffWaysToCompute("2*3-4*5"))

File: c++/src/1-999/440findKthNumber.cpp
#include "lc_pub.h"

class Solution {
public:
    
    int findKthNumber(int n, int k) {
        string sn = to_string(n);
        int m = sn.size();

        // 建立 10 叉树，根节点编号0，后面的数字依次编号，前序遍历此树

        auto calc = [&](int node) -> int {
            // 计算node节点的子节点数量
            string s_node = to_string(node);
            int len = s_node.size();
            int cnt=0;
            if (sn.starts_with(s_node)) {                
                for (int i=0;i<m-len;i++) {
                    cnt=cnt*10+1;
                }
                cnt+=n%int(pow(10,m-len))+1;
            }
            else {
                if (sn.substr(0,len) > s_node) {
                    for (int i=0;i<m-len+1;i++) {
                        cnt=cnt*10+1;
                    }
                }
                else {
                    for (int i=0;i<m-len;i++) {
                        cnt=cnt*10+1;
                    }
                }
                
            }
            return cnt;
        };

        int ans = 0;

        int node = 1;
        while (k > 0) {
            int v = calc(node);
            if (v < k) {
                node ++;  // 找兄弟节点
                k-=v;
            }
            else {
                if (k==1) return node;
                k--;
                node *= 10;  // 找子节点
            }
        }
        return node;
    }
};

int main()
{
    std::cout<<"test let us start! %s" << __cplusplus <<std::endl;
    vector<int> nums{1,2,3};
    Solution so;
    auto v = so.findKthNumber(100, 10);
    cout << v << endl;
    return 0;
}
